packages/multicast_dns/lib/src/packet.dart

```diff
@@ -0,0 +1,373 @@
+/// Result of reading a Fully Qualified Domain Name (FQDN).
+class _FQDNReadResult {
+  /// Creates a new FQDN read result.
+  _FQDNReadResult(this.fqdnParts, this.bytesRead);
+
+  /// The raw parts of the FQDN.
+  final List<String> fqdnParts;
+
+  /// The bytes consumed from the packet for this FQDN.
+  final int bytesRead;
+
+  /// Returns the Fully Qualified Domain Name.
+  String get fqdn => fqdnParts.join('.');
+
+  @override
+  String toString() => fqdn;
+}
+
+/// Reads a FQDN from raw packet data.
+String readFQDN(List<int> packet, [int offset = 0]) {
+  final Uint8List data =
+      packet is Uint8List ? packet : Uint8List.fromList(packet);
+  final ByteData byteData = ByteData.view(data.buffer);
+
+  return _readFQDN(data, byteData, offset, data.length).fqdn;
+}
+
+// Read a FQDN at the given offset. Returns a pair with the FQDN
+// parts and the number of bytes consumed.
+//
+// If decoding fails (e.g. due to an invalid packet) `null` is returned.
+_FQDNReadResult _readFQDN(
+    Uint8List data, ByteData byteData, int offset, int length) {
+  void checkLength(int required) {
+    if (length < required) {
+      throw MDnsDecodeException(required);
+    }
+  }
+
+  final List<String> parts = <String>[];
+  final int prevOffset = offset;
+  while (true) {
+    // At least one byte is required.
+    checkLength(offset + 1);
+
+    // Check for compressed.
+    if (data[offset] & 0xc0 == 0xc0) {
+      // At least two bytes are required for a compressed FQDN.
+      checkLength(offset + 2);
+
+      // A compressed FQDN has a new offset in the lower 14 bits.
+      final _FQDNReadResult result = _readFQDN(
+          data, byteData, byteData.getUint16(offset) & ~0xc000, length);
+      parts.addAll(result.fqdnParts);
+      offset += 2;
+      break;
+    } else {
+      // A normal FQDN part has a length and a UTF-8 encoded name
+      // part. If the length is 0 this is the end of the FQDN.
+      final int partLength = data[offset];
+      offset++;
+      if (partLength > 0) {
+        checkLength(offset + partLength);
+        final Uint8List partBytes =
+            Uint8List.view(data.buffer, offset, partLength);
+        offset += partLength;
+        parts.add(utf8.decode(partBytes));
+      } else {
+        break;
+      }
+    }
+  }
+  return _FQDNReadResult(parts, offset - prevOffset);
+}
+
+/// Decode an mDNS query packet.
+///
+/// If decoding fails (e.g. due to an invalid packet), `null` is returned.
+///
+/// See https://tools.ietf.org/html/rfc1035 for format.
+ResourceRecordQuery decodeMDnsQuery(List<int> packet) {
+  final int length = packet.length;
+  if (length < _kHeaderSize) {
+    return null;
+  }
+
+  final Uint8List data =
+      packet is Uint8List ? packet : Uint8List.fromList(packet);
+  final ByteData packetBytes = ByteData.view(data.buffer);
+
+  // Check whether it's a query.
+  final int flags = packetBytes.getUint16(_kFlagsOffset);
+  if (flags != 0) {
+    return null;
+  }
+  // Number of questions.
```

Comment:
Rather than this comment, I'd just change the name of `qdcount` to something more descriptive: this comment doesn't say much.